bpf/lib/nodeport.h

```diff
@@ -500,6 +499,98 @@ static __always_inline int nodeport_nat_ipv4_rev(struct __sk_buff *skb,
 	return NODEPORT_DO_NAT_IPV4(addr, NAT_DIR_INGRESS);
 }
 
+/* Helper function to set the IPv4 option for DSR when a backend is remote.
+ * NOTE: Revalidate data after calling the function.
+ */
+static inline int set_dsr_opt4(struct __sk_buff *skb, struct iphdr *ip4,
+			       __be32 svc_addr, __be32 svc_port)
+{
+	union tcp_flags tcp_flags = { .value = 0 };
+	__be32 sum;
+	__u32 iph_old, iph_new;
+
+	if (ip4->protocol == IPPROTO_TCP) {
+		if (skb_load_bytes(skb, ETH_HLEN + sizeof(*ip4) + 12,
+				   &tcp_flags, 2) < 0)
+			return DROP_CT_INVALID_HDR;
+		// Setting the option is required only for the first packet
+		// (SYN), in the case of TCP, as for further packets of the
+		// same connection a remote node will use a NAT entry to
+		// reverse xlate a reply.
+		if (!(tcp_flags.value & (TCP_FLAG_SYN)))
+			return 0;
+	}
+
+	iph_old = *(__u32 *)ip4;
+	ip4->ihl += 0x2; // To accommodate u64 option
+	ip4->tot_len = bpf_htons(bpf_ntohs(ip4->tot_len) + 0x8);
+	iph_new = *(__u32 *)ip4;
+	__u32 opt1 = bpf_htonl(DSR_IPV4_OPT_32 | svc_port);
+	__u32 opt2 = bpf_htonl(svc_addr);
+
+	sum = csum_diff(&iph_old, 4, &iph_new, 4, 0);
+	sum = csum_diff(NULL, 0, &opt1, sizeof(opt1), sum);
+	sum = csum_diff(NULL, 0, &opt2, sizeof(opt2), sum);
+
+	if (skb_adjust_room(skb, 0x8, BPF_ADJ_ROOM_NET, 0))
+		return DROP_INVALID;
+	if (skb_store_bytes(skb, ETH_HLEN + sizeof(*ip4),
+			    &opt1, sizeof(opt1), 0) < 0)
+		return DROP_INVALID;
+	if (skb_store_bytes(skb, ETH_HLEN + sizeof(*ip4) + sizeof(opt1),
+			    &opt2, sizeof(opt2), 0) < 0)
+		return DROP_INVALID;
+
+	if (l3_csum_replace(skb, ETH_HLEN + offsetof(struct iphdr, check),
+	    0, sum, 0) < 0) {
+		return DROP_CSUM_L3;
+	}
```

Comment:
Did you check what happens under GSO/TSO? Meaning, large, GRO-aggregated train of skbs comes in, creates a bit one, we adjust room and expect GSO/TSO later on to split it again. Does this work and takes these options into account (I'm currently unsure, but worth double checking the GSO engine code)?

Reply:
Discussed offline: as we set the option only for the first packet, the GRO and GSO/TSO implications are not worrisome.